src/main/java/net/dries007/tfc/compat/wailatop/WailaBlocks.java

```diff
@@ -0,0 +1,502 @@
+        registrar.registerBodyProvider(dataProvider, TEFirePit.class);
+        registrar.registerBodyProvider(dataProvider, TEBloomery.class);
+        registrar.registerBodyProvider(dataProvider, TEBlastFurnace.class);
+        registrar.registerBodyProvider(dataProvider, TECharcoalForge.class);
+        registrar.registerBodyProvider(dataProvider, TEBarrel.class);
+        registrar.registerBodyProvider(dataProvider, TELogPile.class);
+        registrar.registerBodyProvider(dataProvider, TEPlacedItemFlat.class);
+
+        registrar.registerBodyProvider(dataProvider, TECropBase.class);
+        registrar.registerBodyProvider(dataProvider, BlockFarmlandTFC.class);
+        registrar.registerBodyProvider(dataProvider, BlockFruitTreeLeaves.class);
+        registrar.registerBodyProvider(dataProvider, BlockBerryBush.class);
+        registrar.registerBodyProvider(dataProvider, BlockCropDead.class);
+
+        //Tail
+        registrar.registerTailProvider(dataProvider, BlockRockVariant.class);
+
+
+    }
+
+    @Override
+    public ItemStack getWailaStack(IWailaDataAccessor accessor, IWailaConfigHandler config)
+    {
+        Block b = accessor.getBlock();
+        ItemStack itemstack = ItemStack.EMPTY;
+
+        if (b instanceof BlockOreTFC)
+        {
+            itemstack = OreTFCStack(accessor, config);
+        }
+        else if (b instanceof BlockCropSimple)
+        {
+            itemstack = CropSimpleStack(accessor, config);
+        }
+        return itemstack;
+    }
+
+    @Override
+    public List<String> getWailaHead(ItemStack itemStack, List<String> currenttip, IWailaDataAccessor accessor, IWailaConfigHandler config)
+    {
+        Block b = accessor.getBlock();
+        TileEntity te = accessor.getTileEntity();
+
+        if (b instanceof BlockOreTFC) currenttip = OreTFCHead(itemStack, currenttip, accessor, config);
+        if (b instanceof BlockBarrel) currenttip = BarrelHead(itemStack, currenttip, accessor, config);
+        if (te instanceof TECropBase) currenttip = CropSimpleHead(itemStack, currenttip, accessor, config);
+        if (b instanceof BlockCropDead) currenttip = CropDeadHead(itemStack, currenttip, accessor, config);
+        if (b instanceof BlockFruitTreeTrunk | b instanceof BlockFruitTreeLeaves | b instanceof BlockFruitTreeBranch)
+        {
+            currenttip = FruitTreeHead(itemStack, currenttip, accessor, config);
+        }
+        return currenttip;
+    }
+
+    @Override
+    public List<String> getWailaBody(ItemStack itemStack, List<String> currenttip, IWailaDataAccessor accessor, IWailaConfigHandler config)
+    {
+        Block b = accessor.getBlock();
+        TileEntity te = accessor.getTileEntity();
+
+        // Mechanics
+        if (b instanceof BlockOreTFC) currenttip = OreTFCBody(itemStack, currenttip, accessor, config);
+        else if (te instanceof TEPlacedItemFlat)
+            currenttip = PlacedItemFlatBody(itemStack, currenttip, accessor, config);
```

Comment:
Please expand this:
```suggestion
        if (b instanceof BlockOreTFC) 
        {
            return OreTFCBody(itemStack, currenttip, accessor, config);
        }
        else if (te instanceof TEPlacedItemFlat)
        {
            return PlacedItemFlatBody(itemStack, currenttip, accessor, config);
        }
```
etc.

I'm not a fan of this giant if-else statement but if we were to change it I'm not sure it would be much better. So it stays.

Reply:
Yeah I looked at a number of other methods to resolve. Maps and enums and various OO pretty but awkward solutions.. In the end it really is the marginally best of a set of choices.